chore: Installation script improvements

mirkosoft opened the pull request:
Installation script has been moved to the main monorepo. This PR also contains improvements to it so script won't need any manual interactions when run.

closes #7375

mirkosoft committed:
installation script v2 added

mirkosoft committed:
update versions

mirkosoft committed:
configurable base dir
bug fixes

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

hackergil reviewed (changes requested):
Getting the following error in Unix:

+-----------------------------+
| Updating projects versions. |
+-----------------------------+

core
/Users/i867028/git/github/spartacus-latest/clone/projects/core
sed: illegal option -- -
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]

Also, I don't think we need the windows support. Feel free to get rid of it unless Divante team will use it.

marlass commented on scripts/install/config.default.sh:
```diff
@@ -0,0 +1,35 @@
+        "schematics"
+        )
+
+SPARTACUS_REPO_URL="git://github.com/SAP/cloud-commerce-spartacus-storefront.git"
```
Let's update this to new spartacus repo url

marlass commented on scripts/install/README.md:
```diff
@@ -0,0 +1,93 @@
+# Spartacus Installation Script
```
Could we also update the docs in this PR?

marlass commented on scripts/.gitignore:
```diff
@@ -1,3 +1,13 @@
 *.d.ts
+
+#
+# Output from the installation script
```
Do we need it as we create the directory outside of the repo?

marlass replied:
at least the clone/apps

Commit:
sed fix
docs update

mirkosoft commented:
> Getting the following error in Unix:
> 
> +-----------------------------+
> | Updating projects versions. |
> +-----------------------------+
> 
> core
> /Users/i867028/git/github/spartacus-latest/clone/projects/core
> sed: illegal option -- -
> usage: sed script [-Ealn] [-i extension] [file ...]
> sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
> 
> Also, I don't think we need the windows support. Feel free to get rid of it unless Divante team will use it.

Let's leave the windows support as most of Divante devs are using this OS. Removed the --debug option from sed.

mirkosoft committed:
sed fix

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

mirkosoft committed:
Update readme

mirkosoft committed:
Merge branch 'feature/GH-7375_installation_script' of https://github.com/SAP/spartacus into feature/GH-7375_installation_script

mirkosoft committed:
update README

marlass commented on scripts/install/README.md:
````diff
@@ -0,0 +1,85 @@
+```
+so it's one level above the spartacus project root. To change the output directory `$BASE_DIR` can be overriden in the `config.sh` file.
+
+All scirpt's commands should be run while being in it's root directory.
````
```suggestion
All script's commands should be run while being in it's root directory.
```

marlass commented on scripts/install/README.md:
````diff
@@ -0,0 +1,85 @@
+./run.sh start
+```
+
+This command starts a development server for Spartacus, which is under `"${BASE_DIR}/apps/csr"` by default.
````
```suggestion
This command starts a server for Spartacus, which is under `"${BASE_DIR}/apps/csr"` by default.
```

marlass commented on scripts/install/README.md:
````diff
@@ -0,0 +1,85 @@
+./run.sh stop
+```
+
+This command stops the running Spartacus using its PID.
````
```suggestion
This command stops the running Spartacus PM2 processes.
```

marlass commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,329 @@
+    for i in $projects
+        do
+            echo $i;
+            (cd "${CLONE_DIR}/projects/${i}" && pwd && sed -i -E 's/"version": "[^"]+/"version": "'"${SPARTACUS_VERSION}"'/g' package.json);
```
Can we fix the case when someone passes `latest` or `next` here in local install? Then it won't work. Publish will fail with those 2. We could set it maybe for very high version `999.999.999`? What you think?

mirkosoft replied:
What's the use case for passing `latest` or `next` in local install?

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

Commit:
Apply suggestions from code review

Co-authored-by: Marcin Lasak <lasakmarcin90@gmail.com>

Commit:
set fixed version for verdaccio
add check for verdaccio

marlass commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,337 @@
+        echo "Directory ${BASE_DIR} already exists, please remove it in order to start fresh installation."
+        exit 1
+    fi
+    VERDACCIO_PID=`lsof -nP -i4TCP:4873 | grep LISTEN | tr -s ' ' | cut -d ' ' -f 2`
```
On Windows it doesn't work, but also doesn't terminate the job, so it can stay.

marlass commented on scripts/.gitignore:
```diff
@@ -1,3 +0,0 @@
-*.js
```
Could we bring this file back? I think it's required for tests and builds

mirkosoft committed:
bring back scripts/.gitignore

mirkosoft committed:
Merge branch 'feature/GH-7375_installation_script' of https://github.com/SAP/spartacus into feature/GH-7375_installation_script

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

hackergil commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,337 @@
+}
+
+function delete_dir {
+    if [ -d "${1}" ]; then
```
Assigning the first argument to a variable makes the function easier to change if needed

hackergil commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,337 @@
+    fi
+    VERDACCIO_PID=`lsof -nP -i4TCP:4873 | grep LISTEN | tr -s ' ' | cut -d ' ' -f 2`
+    if [[ -n ${VERDACCIO_PID} ]]; then
+        echo "It seems verdaccio is already running on PID: ${VERDACCIO_PID}. Please kill it first to continue."
```
Is there a reason why this can happen? Is it common?

mirkosoft replied:
If something goes wrong between starting verdaccio and killing it, next run of the script used to have problems with already running verdaccio (with some of libs already published). What do you think of automatically killing it instead of exit?

giancorderoortiz replied:
It happens in dev-servers sometimes.

marlass replied:
I would kill it

hackergil commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,337 @@
+
+# top directory for the installation output (must be outside of the project)
+if [ -z $BASE_DIR ]; then
+    BASE_DIR="../../../spartacus-${SPARTACUS_VERSION}"
```
can we extract this to a variable at the top? I am guessing we might change it soon.

mirkosoft replied:
Actually, it's intended to use config file to for this. If `BASE_DIR` is set in the `config.sh` - this takes the precedence. The reason of having the variable not at the top is it has to be after config resolution (lines above).

hackergil reviewed (changes requested):
* I get a prompt for npm password on each lib build. Is there a way to bypass the password request on each package? If so, can we add it to the README?
* Can we get rid of "Directory ../../../spartacus-latest already exists, please remove it in order to start fresh installation."? When we automate this (for CD for example) we'll need the script to just reinstall without asking
* I am getting the following error on mac:
http <-- 200, user: null(127.0.0.1), req: 'GET /@spartacus%2fschematics/-/schematics-999.999.999.tgz', bytes: 0/131086
Installed packages for tooling via npm.
            ✅️ Added '@spartacus/core' into dependencies
            ✅️ Added '@spartacus/storefront' into dependencies
            ✅️ Added '@spartacus/assets' into dependencies
            ✅️ Added '@spartacus/styles' into dependencies
            ✅️ Added '@ng-bootstrap/ng-bootstrap' into dependencies
            ✅️ Added '@ng-select/ng-select' into dependencies
            ✅️ Added '@ngrx/store' into dependencies
            ✅️ Added '@ngrx/effects' into dependencies
            ✅️ Added '@ngrx/router-store' into dependencies
            ✅️ Added 'bootstrap' into dependencies
            ✅️ Added 'i18next' into dependencies
            ✅️ Added 'i18next-xhr-backend' into dependencies
            ✅️ Added '@angular/service-worker' into dependencies
            ✅️ Added '@angular/localize' into dependencies
            ✅️ Added 'ngx-infinite-scroll' into dependencies
Cannot read property 'kind' of undefined

mirkosoft commented:
> * I get a prompt for npm password on each lib build. Is there a way to bypass the password request on each package? If so, can we add it to the README?

Yes, I just run into it when trying one of the servers. You have to add `//localhost:4873/:_authToken="VNOlucnBmPrXVjzHVEHtWQ=="` to your `.npmrc` to workaround this problem (it might be a fake token). Will update README.

> * Can we get rid of "Directory ../../../spartacus-latest already exists, please remove it in order to start fresh installation."? When we automate this (for CD for example) we'll need the script to just reinstall without asking

Sure, already asked by GC, will change that.
> * I am getting the following error on mac:
>   http <-- 200, user: null(127.0.0.1), req: 'GET /@spartacus%2fschematics/-/schematics-999.999.999.tgz', bytes: 0/131086
>   Installed packages for tooling via npm.
>   white_check_mark Added '@spartacus/core' into dependencies
>   white_check_mark Added '@spartacus/storefront' into dependencies
>   white_check_mark Added '@spartacus/assets' into dependencies
>   white_check_mark Added '@spartacus/styles' into dependencies
>   white_check_mark Added '@ng-bootstrap/ng-bootstrap' into dependencies
>   white_check_mark Added '@ng-select/ng-select' into dependencies
>   white_check_mark Added '@ngrx/store' into dependencies
>   white_check_mark Added '@ngrx/effects' into dependencies
>   white_check_mark Added '@ngrx/router-store' into dependencies
>   white_check_mark Added 'bootstrap' into dependencies
>   white_check_mark Added 'i18next' into dependencies
>   white_check_mark Added 'i18next-xhr-backend' into dependencies
>   white_check_mark Added '@angular/service-worker' into dependencies
>   white_check_mark Added '@angular/localize' into dependencies
>   white_check_mark Added 'ngx-infinite-scroll' into dependencies
>   Cannot read property 'kind' of undefined

This is caused by the wrong version of `ng` (app was created by `9.0.7`). It must match the requirements  we have in the project's `package.json` (currently it's `~9.1.0`). Worth to mention in the README as well, I guess.

mirkosoft committed:
changes reqested

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

Commit:
Update ng cli version

hackergil reviewed (changes requested):
I reviewed your changes and they look good. Thank you.

I was able to run the script but when I start the server I get an error in the browser (both Chrome and Firefox)

main-es2015.784066ea6e80ff10cdbb.js:1 Service worker registration failed with: DOMException: Failed to register a ServiceWorker for scope ('http://localhost:4200/') with script ('http://localhost:4200/ngsw-worker.js'): The script has an unsupported MIME type ('text/html').

Also, what does `csr` mean? :)

mirkosoft commented:
> I reviewed your changes and they look good. Thank you.
> 
> I was able to run the script but when I start the server I get an error in the browser (both Chrome and Firefox)
> 
> main-es2015.784066ea6e80ff10cdbb.js:1 Service worker registration failed with: DOMException: Failed to register a ServiceWorker for scope ('http://localhost:4200/') with script ('http://localhost:4200/ngsw-worker.js'): The script has an unsupported MIME type ('text/html').
> 
> Also, what does `csr` mean? :)

Error is caused by not running ssr app or an issue related to ssr serving wrong mime type for service worker (we use `serve`, needs to be checked). Csr stands for "client side rendering" :) If you have better name let's change it, sure!

mirkosoft committed:
add ssr, minor fixes

mirkosoft committed:
add missing OCC_PREFIX

mirkosoft committed:
add ssr_pwa app
remove not needed pwa app
turn off ng analytics
increase wait time for verdaccio

mirkosoft committed:
mv ng analytics off to better place

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script

hackergil committed:
Merge branch 'develop' into feature/GH-7375_installation_script

hackergil committed:
GH-7375 Minor changes and updating README

mirkosoft committed:
add currency support
enhance pm2 process names
fix version altering

mirkosoft committed:
fix stop_apps

mirkosoft committed:
update new-release issue template

marlass commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,343 @@
+    printf "\033[0m" # end green color
+}
+
+function integrate_with_smartedit {
```
Can we create a issue to integrate smartedit later?

marlass commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,343 @@
+    create_app ssr
+}
+
+function create_pwa {
```
This function can be removed

marlass commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,343 @@
+    else
+        prestart_ssr
+        printh "Starting ssr app"
+        ( cd ${INSTALLATION_DIR}/ssr && export PORT=${SSR_PORT} && pm2 start --name "${BASE_SITE}-ssr-${SSR_PORT}" dist/ssr/server/main.js )
```
This SSR works without the NODE_TLS.. flag?

mirkosoft replied:
I guess it shouldn't. Will add it here as well.

marlass commented on scripts/install/config.default.sh:
```diff
@@ -0,0 +1,40 @@
+        "assets"
+        "storefrontlib"
+        "storefrontstyles"
+        "cds"
```
I think we can skip CDS here now and create ticket to allow app instalation with CDS (with dev schematics)

mirkosoft replied:
OK

marlass commented on scripts/install/README.md:
````diff
@@ -0,0 +1,101 @@
+
+```bash
+./run.sh stop
+./run.sh install (script ask you to rename/delete the `$BASE_DIR` if it exists)
````
It will ask or it just removes the directory?

mirkosoft replied:
It will remove, it's the old message.

marlass commented on scripts/install/run.sh:
```diff
@@ -0,0 +1,343 @@
+    else
+        prestart_csr
+        printh "Starting csr app"
+        pm2 start --name "${BASE_SITE}-csr-${CSR_PORT}" serve -- ${INSTALLATION_DIR}/csr/dist/csr/ --single -p ${CSR_PORT}
```
Do we need to use base sites here?

mirkosoft replied:
Added it as if we run multiple basesites apps names alone clashes together. It was enough to use just PORT but with base_site you know immediately which app is which when looking at pm2 processes.

marlass replied:
Ok. Probably we could remove that, when we will use dev-schematics and we will have all base sites available under single application.

mirkosoft replied:
Yes, exactly.

mirkosoft committed:
requested changes

marlass committed:
Merge branch 'develop' into feature/GH-7375_installation_script

Commit:
Merge branch 'develop' into feature/GH-7375_installation_script
